index.rst

```diff
@@ -338,4 +338,121 @@ smaller than a single alert message.
+there is a significant impact on Kafka compute utilization and
+timing when postage stamp cutouts are included in the alert packets.  Additional consideration should be
+given to alternatives for the distribution of stamps, if possible.  One alternative to including postage
+stamps in the stream, recommended by other users of Kafka for large messages, is to forego including the
+stamps in the stream and instead include a url to the location of the stamps from which the stamp can be
+accessed separately.  This would particularly alleviate the increased network traffic for each additional
+consumer reading a stream.  Another potentially possible alternative could be to separate the postage stamps
+into their own Kafka topic.  If not all primary consumers are required or expected to read postage stamp
+cutouts and pass them to downstream users, this could decrease the total network traffic, since not all
+consumers would pull from the larger stream of stamps.  A downstream filter could then unite alerts of interest
+with corresponding stamps.  The feasibility of this option for combining streams still needs to be further investigated.
+
+It is unclear why an increased number of consumers would significantly slow a single producer as to be unable
+to keep up with the 39 second separation between bursts.  With larger instance sizes, there should be no bottleneck
+in reading from disk and no bottleneck in the total network traffic.  Though the parallelization of compute per CCD, i.e.,
+an increased number of producers in parallel, should eliminate this problem, there is the possibility of
+performance improvement if this issue can be addressed.  Preliminary testing indicates that adjusting three
+parameters, ``buffer.memory``, ``batch.size``, and ``linger.ms``, together may help.  Using multiple partitions
+may also help.  All tests above use a single partition for a single topic.
+
+It is also unclear why the consumers consumption time is slowed for the case of 10+1 consumers reading a stream
+with stamps.  The consumer consumption time is outside the total alert production pipeline time for the alerts
+to be accessible from Kafka, and a slow consumer can have additional processing while reading the stream that
+can slow the process, but the transit and consumption time should still be considered.  Consumer reading of the
+stream can easily be parallelized for faster processing.  The parallelization is controlled by the number of
+topic partitions, so increasing the number of partitions can decrease consumer read time such that consumers
+do not lag behind bursts if the consumers read in consumer groups in parallel.  For the critical time from
+producer serialization into Avro and submission to Kafka, the time it takes for the alert distribution to get
+alerts into the queue and ready to be read by consumers can be expected to be about 6-9 seconds when stamps
+are included and 1-3 seconds without stamps.
```

Comment:
Again, this speaks to my concern above. I *think* what I'm learning here is that the consumers read sequentially (round-robin?) from a single partition, in which case it makes sense that things slow down with many consumers. If the parallelization is controlled by the partitions, then it seems like this would be an excellent thing to investigate.

Reply:
Agreed, yes.